sympy/physics/quantum/represent.py

```diff
@@ -171,7 +171,14 @@ def _represent_FooBasis(self, e, basis, **options)
         base, exp = expr.as_base_exp()
         if format == 'numpy' or format == 'scipy.sparse':
             exp = _sympy_to_scalar(exp)
-        return represent(base, **options)**exp
+        base = represent(base, **options)
+        # scipy.sparse doesn't support negative exponents
+        # and warns when inverting a matrix in csr format.
```

Comment:
From the code, it looks like either format would do.
https://github.com/scipy/scipy/blob/81d2318e3a9ab172c05645e5d663979f7c594472/scipy/sparse/linalg/matfuncs.py#L77-L82
https://github.com/scipy/scipy/blob/81d2318e3a9ab172c05645e5d663979f7c594472/scipy/sparse/linalg/dsolve/linsolve.py#L135-L138

Reply:
The warnings come from these two places:
https://github.com/scipy/scipy/blob/81d2318e3a9ab172c05645e5d663979f7c594472/scipy/sparse/linalg/dsolve/linsolve.py#L308-L310
https://github.com/scipy/scipy/blob/81d2318e3a9ab172c05645e5d663979f7c594472/scipy/sparse/linalg/dsolve/linsolve.py#L206-L209

Reply:
I see, tocsc has to be used. But could we dispense with tocsr?

Reply:
I guess I was trying to change as little as possible. Do you have any idea why csr format is used at all here?

Among the tests the case for which this code is needed is for a 2x2 matrix so using any sparse format is unnecessary.

Reply:
> Do you have any idea why csr format is used at all here?

I suspect that it was chosen by chance. Perhaps this could be changed as scipy seems to prefer csc: https://github.com/sympy/sympy/blob/029936987c7b35fcc6b347b580d4c154792344e7/sympy/physics/quantum/matrixutils.py#L70

Reply:
There are quite a few references to csr format around:
```console
$ git grep csr sympy/physics/quantum/
sympy/physics/quantum/matrixcache.py:    ``scipy.sparse.csr_matrix`` matrices. These matrices are then stored for
sympy/physics/quantum/matrixutils.py:        return sparse.csr_matrix(np.matrix(m.tolist(), dtype=dtype))
sympy/physics/quantum/matrixutils.py:        return sparse.csr_matrix(m)
sympy/physics/quantum/matrixutils.py:    # The final matrices will just be multiplied, so csr is a good final
sympy/physics/quantum/matrixutils.py:    return sparse.csr_matrix(answer)
sympy/physics/quantum/matrixutils.py:    spmatrix = options.get('spmatrix', 'csr')
sympy/physics/quantum/matrixutils.py:    elif spmatrix == 'csr':
sympy/physics/quantum/matrixutils.py:        return sparse.csr_matrix((m, n), dtype=dtype)
sympy/physics/quantum/qubit.py:            return sparse.csr_matrix(result, dtype='complex').transpose()
sympy/physics/quantum/represent.py:        # and warns when inverting a matrix in csr format.
sympy/physics/quantum/represent.py:            base = inv(base.tocsc()).tocsr()
sympy/physics/quantum/sho1d.py:            matrix = matrix.tocsr()
sympy/physics/quantum/sho1d.py:            matrix = matrix.tocsr()
sympy/physics/quantum/sho1d.py:            matrix = matrix.tocsr()
sympy/physics/quantum/sho1d.py:            matrix = matrix.tocsr()
sympy/physics/quantum/sho1d.py:                vector = vector.tocsr()
sympy/physics/quantum/sho1d.py:                vector = vector.tocsr()
sympy/physics/quantum/tests/test_dagger.py:    a = sparse.csr_matrix([[1.0 + 0.0j, 2.0j], [-1.0j, 2.0 + 0.0j]])
sympy/physics/quantum/tests/test_matrixutils.py:    result = sparse.csr_matrix([[1, 2], [3, 4]], dtype='complex')
```
This is the only explanation I see:
https://github.com/sympy/sympy/blob/f73034eaa8342dc70dc8b27da7cbbce0ee9e5e95/sympy/physics/quantum/matrixutils.py#L172-L174
https://github.com/sympy/sympy/blob/f73034eaa8342dc70dc8b27da7cbbce0ee9e5e95/sympy/physics/quantum/matrixcache.py#L76-L80

I think that in general the inverse of a sparse matrix is usually not sparse anyway so the operation here is unlikely to have any kind of sparse efficiency.

Reply:
>  The final matrices will just be multiplied, so csr is a good final sparse format. 

Well, it seems that there was some reason for the choice though I expect that csc (or even dense) would work as well. Anyway, as csr is used in so many places, let's leave it as it is. I think this is good to go.